core/amber/src/main/python/core/models/RTupleExecutor.py

```diff
@@ -0,0 +1,193 @@
+        if isinstance(value, robjects.vectors.POSIXct):
+            return next(value.iter_localized_datetime())
+        else:
+            return float(value[0])
+    if isinstance(value, robjects.vectors.StrVector):
+        return str(value[0])
+    return value
+
+
+class RTupleExecutor(TupleOperatorV2):
+    """
+    An operator that can execute R code on R Lists (R's representation of a Tuple)
+    """
+
+    is_source = False
+
+    _combine_binary_and_non_binary_lists = robjects.r(
+        """
+        function(non_binary_list, binary_list) {
+            non_binary_list <- as.list(non_binary_list$as_vector())
+            return (c(non_binary_list, binary_list))
+        }
+        """
+    )
+
+    def __init__(self, r_code: str):
+        """
+        Initialize the RTupleExecutor with R code.
+
+        Args:
+            r_code (str): R code to be executed.
+        """
+        super().__init__()
+        # Use the local converter from rpy2 to load in the R function given by the user
+        with local_converter(default_converter):
+            self._func = robjects.r(r_code)
+
+    def process_tuple(self, tuple_: Tuple, port: int) -> Iterator[Optional[TupleLike]]:
+        """
+        Process an input Tuple from the given link.
+
+        :param tuple_: Tuple, a Tuple from an input port to be processed.
+        :param port: int, input port index of the current Tuple.
+        :return: Iterator[Optional[TupleLike]], producing one TupleLike object at a
+            time, or None.
+        """
+        with local_converter(arrow_converter):
+            input_schema: pa.Schema = tuple_._schema.as_arrow_schema()
+            input_fields: list[str] = [field.name for field in input_schema]
+            non_binary_fields: list[str] = [
+                field.name for field in input_schema if field.type != pa.binary()
+            ]
+            binary_fields: list[str] = [
+                field.name for field in input_schema if field.type == pa.binary()
+            ]
+
+            non_binary_tuple: Tuple = tuple_.get_partial_tuple(non_binary_fields)
+            non_binary_tuple_schema: pa.Schema = (
+                non_binary_tuple._schema.as_arrow_schema()
+            )
+            non_binary_pyarrow_array: pa.StructArray = pa.array(
+                [non_binary_tuple.as_dict()], type=pa.struct(non_binary_tuple_schema)
+            )
+
+            binary_tuple: Tuple = tuple_.get_partial_tuple(binary_fields)
+            binary_r_list: dict[str, object] = {}
+            for k, v in binary_tuple.as_dict().items():
+                if isinstance(v, bytes):
+                    binary_r_list[k] = pickle.loads(v[10:])
+                elif isinstance(v, datetime.datetime):
+                    binary_r_list[k] = robjects.vectors.POSIXct.sexp_from_datetime([v])
+                else:
+                    binary_r_list[k] = v
+
+            binary_r_list: rpy2.robjects.ListVector = robjects.vectors.ListVector(
```

Comment:
name it `binary_r_pyarrow_array`?

Reply:
I name it binary_r_list because it is directly translated into an rpy2 ListVector, which is then translated to R. Only the columns with non-binary fields are turned into a pyarrow StructArray.